### ISSUE TITLE
Inline images and attachments for texts in storage system

### ISSUE THREAD

**Issue by bedeho** [Joystream/pioneer#55]
There are lots of areas in Pioneer where one has to compose a rich markdown text, this includes

- proposal texts
- forum posts
- canidacy profiles
- opening descriptions
- applications
- etc.

In all of these places it should be possible to embed images and separate attachments that all are stored to and, when displayed, loaded from the built in Joystream storage system. See this issue for product focused description of storage system.

Joystream/joystream#2104

This will require more sophisticated composer and rendering experience for such text fields where we currently just have a generic markdown editor.

**Comment by jodator** [comment]
> This will require more sophisticated composer and rendering experience for such text fields where we currently just have a generic markdown editor.

In my previous role, I was a core developer of the CKEditor 5 project and I can recommend it.

The CKEditor 5 can be configured to be a markdown editor while offering a text processor experience. (I'm actually using [GithubWriter plugin](https://github.com/ckeditor/github-writer) to write longer GH issues descriptions).

The [image upload can be customized](https://ckeditor.com/docs/ckeditor5/latest/features/image-upload/image-upload.html) as well and is [licensed](https://github.com/ckeditor/ckeditor5/blob/master/LICENSE.md) under the GPLv2.

**Comment by bedeho** [comment]
That sounds interesting, could the editor be reworked to support this also?

https://github.com/Joystream/pioneer/issues/50

**Comment by jodator** [comment]
> That sounds interesting, could the editor be reworked to support this also?
> 
> #50

I think that in short: yes, I'll post details there.

**Comment by traumschule** [comment]
+1

**Comment by bedeho** [comment]
That was a stopgap, since there was no on-chain integration between membership pallet and storage system, really not good to try to add anything more than that, its no better than using imgur really. Our own storage system for our own assets is what we really should stick to, and for non-urgent deliveries we should try to do it right.

**Comment by kdembler** [comment]
Agreed with the above!

**Comment by traumschule** [comment]
So to be able to use pioneer to upload images Pioneer would need to have ownership for a dedicated channel. Since it's a FE app that can be deployed by different operators each would need to own a channel for their respective instance. The Pioneer BE could be modified to handle uploads.
Another (much easier) option is to require the author to own and designate a channel to upload images to. Pioneer could offer to create one if needed. Thus staying in control (to delete uploads later) and responsible for content. 
Either way would involve duplicating logic from atlas studio.

The third possibility is to have the DAO deploy and maintain a BE service that allows for easy image uploads (to their channel) for other apps  (eg. drag and drop in Pioneer modal), via POST request returning the image URL which could then be embedded into the post by Pioneer. With proper app authentication.

**Comment by bedeho** [comment]
> So to be able to use pioneer to upload images Pioneer would need to have ownership for a dedicated channel.

No, there are membership bags which are meant to encompass member level storage, and also working group and council bags, for same reason. Just not enabled I believe, runtime integration between membership and council pallets and storage pallet.
